### PR TITLE
[RFC] Continue generating examples after we find the first bug 

### PR DESCRIPTION
**Pull request by DRMacIver**
Attempt at #847. Implements the "Try up until twice the number of examples it took you to discover the last new bug" heuristic.

It seems to work OK? I struggled a lot to get good testing of this at first. I feel like I want to:

* [ ] Think about the heuristics a bit more
* [ ] Think about the failure modes a lot more
* [ ] See if anything horrendous breaks on the build

Before I commit to this approach. So it might be that this might be actually very close to the final version, or I might end up rewriting the whole thing. Commentary on any or all of the above would also be useful.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Oh, worth noting is that I fudged up the initial run so that if max_examples is >= 200 it always tries at least 200 examples even if the last bug discovery heuristic says it should stop sooner. This prevents you getting unlucky and finding a bug really early on and having it bail out going "Well, that's probably it".

**Comment by Zac-HD** [comment]
I actually think this is pretty much done, with the caveat that I've only skimmed the details.  

The change is conceptually pretty simple - now that the multiple-reporting stuff works, this is just (ha ha) about keeping on looking for examples for "a little while" after finding the first bug.  Too short is no worse than present and too long is at worst as long as shrinking + the no bugs case, so it can't go too badly wrong either way!

    self.valid_examples < min(self.settings.max_examples, max(200, self.last_new_discovery * 2))

I'm going to disregard English because I got the wrong idea from your description - this is both the current code and exact algorithm I was going to suggest 😄 

(all based on if the tests are passing; looks like minimisation is being counted towards the limit and therefore shrinking gets cut off early)

**Comment by DRMacIver** [comment]
> looks like minimisation is being counted towards the limit and therefore shrinking gets cut off early

It's actually not that. What was happening was that because `last_data` was non-interesting if we generated something else after the interesting example, it was triggering flaky test detection.

I think after this pull request I'm going to do a pull request to purge all references to `last_data` from the engine.

File: tests/cover/test_conjecture_engine.py [hunk 1]
```diff
@@ -380,17 +380,16 @@ def f(data):
...
+    max_examples = 10
 
-    runner = ConjectureRunner(f, settings=settings(
+    runner = ConjectureRunner(slow, settings=settings(
         database=None, phases=(Phase.reuse, Phase.generate),
+        max_shrinks=10**6, max_examples=max_examples
     ))
     runner.run()
-    assert len(seen) == 1
+    runner.valid_examples <= max_examples
```
Comment (Zac-HD) [inline tests/cover/test_conjecture_engine.py]: Missing an `assert` here?

Comment (DRMacIver) [reply tests/cover/test_conjecture_engine.py]: Yes, thanks.

**Comment by DRMacIver** [comment]
I'm going to close this for now. I don't have much time to focus on it at the moment, and want to give it a bit of a rethink. I'll open a superseding version at some point later.

**Pull request closed without merging**